File: app/src/main/java/br/com/jonathanzanella/myexpenses/expense/ShowExpenseActivity.java
package br.com.jonathanzanella.myexpenses.expense;

import android.content.Intent;
import android.os.AsyncTask;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.annotation.UiThread;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.TableRow;
import android.widget.TextView;

import br.com.jonathanzanella.myexpenses.R;
import br.com.jonathanzanella.myexpenses.bill.Bill;
import br.com.jonathanzanella.myexpenses.bill.BillRepository;
import br.com.jonathanzanella.myexpenses.chargeable.Chargeable;
import br.com.jonathanzanella.myexpenses.database.RepositoryImpl;
import br.com.jonathanzanella.myexpenses.helpers.CurrencyHelper;
import br.com.jonathanzanella.myexpenses.receipt.Receipt;
import br.com.jonathanzanella.myexpenses.views.BaseActivity;
import butterknife.BindView;

public class ShowExpenseActivity extends BaseActivity implements ExpenseContract.View {
	public static final String KEY_EXPENSE_UUID = ExpensePresenter.KEY_EXPENSE_UUID;

	@BindView(R.id.act_show_expense_name)
	TextView expenseName;
	@BindView(R.id.act_show_expense_date)
	TextView expenseDate;
	@BindView(R.id.act_show_expense_value)
	TextView expenseIncome;
	@BindView(R.id.act_show_expense_value_to_show_in_overview)
	TextView expenseIncomeToShowInOverview;
	@BindView(R.id.act_show_expense_chargeable)
	TextView expenseChargeable;
	@BindView(R.id.act_show_expense_charge_next_month)
	TableRow chargeNextMonth;

	private ExpensePresenter presenter;

	@UiThread
	@Override
	protected void onCreate(@Nullable Bundle savedInstanceState) {
		ExpenseRepository expenseRepository = new ExpenseRepository(new RepositoryImpl<Expense>(this));
		presenter = new ExpensePresenter(expenseRepository, new BillRepository(new RepositoryImpl<Bill>(this), expenseRepository));
		super.onCreate(savedInstanceState);
		setContentView(R.layout.activity_show_expenses);
	}

	@UiThread
	@Override
	protected void onPostCreate(Bundle savedInstanceState) {
		super.onPostCreate(savedInstanceState);
		presenter.onViewUpdated(false);
	}

	@UiThread
	@Override
	protected void storeBundle(Bundle extras) {
		super.storeBundle(extras);

		if(extras != null)
			presenter.storeBundle(extras);
	}

	@Override
	protected void onSaveInstanceState(Bundle outState) {
		super.onSaveInstanceState(outState);
		presenter.onSaveInstanceState(outState);
	}

	@UiThread
	@Override
	protected void onStart() {
		super.onStart();
		presenter.attachView(this);
	}

	@UiThread
	@Override
	protected void onStop() {
		super.onStop();
		presenter.detachView();
	}

	@Override
	public boolean onCreateOptionsMenu(Menu menu) {
		getMenuInflater().inflate(R.menu.edit_delete, menu);
		return true;
	}

	@Override
	public boolean onOptionsItemSelected(MenuItem item) {
		switch (item.getItemId()) {
			case R.id.action_edit: {
				presenter.edit(this);
				break;
			}
			case R.id.action_delete: {
				presenter.delete(this);

				break;
			}
		}
		return super.onOptionsItemSelected(item);
	}

	@UiThread
	@Override
	public void showExpense(final Expense expense) {
		expenseName.setText(expense.getName());
		expenseDate.setText(Receipt.SIMPLE_DATE_FORMAT.format(expense.getDate().toDate()));
		expenseIncome.setText(CurrencyHelper.format(expense.getValue()));
		expenseIncomeToShowInOverview.setText(CurrencyHelper.format(expense.getValueToShowInOverview()));
		new AsyncTask<Void, Void, Chargeable>() {

			@Override
			protected Chargeable doInBackground(Void... voids) {
				return expense.getChargeableFromCache();
			}

			@Override
			protected void onPostExecute(Chargeable chargeable) {
				super.onPostExecute(chargeable);
				expenseChargeable.setText(chargeable.getName());
			}
		}.execute();
		chargeNextMonth.setVisibility(expense.isChargedNextMonth() ? View.VISIBLE : View.GONE);
	}

	@UiThread
	@Override
	protected void onActivityResult(int requestCode, int resultCode, Intent data) {
		super.onActivityResult(requestCode, resultCode, data);
		presenter.attachView(this);
		presenter.onActivityResult(requestCode, resultCode);
	}
}

File: app/src/main/kotlin/br/com/jonathanzanella/myexpenses/expense/ShowExpenseActivity.kt
package br.com.jonathanzanella.myexpenses.expense

import android.content.Intent
import android.os.Bundle
import android.support.annotation.UiThread
import android.support.v7.app.AppCompatActivity
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.TableRow
import android.widget.TextView
import br.com.jonathanzanella.myexpenses.App
import br.com.jonathanzanella.myexpenses.R
import br.com.jonathanzanella.myexpenses.helpers.toCurrencyFormatted
import br.com.jonathanzanella.myexpenses.transaction.Transaction
import br.com.jonathanzanella.myexpenses.views.anko.*
import org.jetbrains.anko.*
import javax.inject.Inject

class ShowExpenseActivity : AppCompatActivity(), ExpenseContract.View {
    override val context = this
    private val ui = ShowExpenseActivityUi()
    @Inject
    lateinit var presenter: ExpensePresenter

    init {
        App.getAppComponent().inject(this)
    }

    @UiThread
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        ui.setContentView(this)

        storeBundle(savedInstanceState)
        storeBundle(intent.extras)

        ui.toolbar.setup(this)
    }

    @UiThread
    override fun onPostCreate(savedInstanceState: Bundle?) {
        super.onPostCreate(savedInstanceState)
        presenter.onViewUpdated(false)
    }

    @UiThread
    fun storeBundle(extras: Bundle?) {
        if (extras != null)
            presenter.storeBundle(extras)
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        presenter.onSaveInstanceState(outState)
    }

    @UiThread
    override fun onStart() {
        super.onStart()
        presenter.attachView(this)
    }

    @UiThread
    override fun onStop() {
        super.onStop()
        presenter.detachView()
    }

    override fun setTitle(string: String) {
        ui.toolbar.title = string
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.edit_delete, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.action_edit -> {
                presenter.edit(this)
            }
            R.id.action_delete -> {
                presenter.delete(this)
            }
        }
        return super.onOptionsItemSelected(item)
    }

    @UiThread
    override fun showExpense(expense: Expense) {
        ui.apply {
            name.text = expense.name
            date.text = Transaction.SIMPLE_DATE_FORMAT.format(expense.getDate().toDate())
            value.text = expense.value.toCurrencyFormatted()
            showInOverview.text = expense.valueToShowInOverview.toCurrencyFormatted()
            chargeNextMonth.visibility = if (expense.chargedNextMonth) View.VISIBLE else View.GONE
        }

        doAsync {
            val chargeable = expense.chargeableFromCache

            uiThread { ui.chargeable.text = chargeable?.name }
        }
    }

    @UiThread
    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        presenter.attachView(this)
        presenter.onActivityResult(requestCode, resultCode)
    }

    companion object {
        val KEY_EXPENSE_UUID = ExpensePresenter.KEY_EXPENSE_UUID
    }
}

private class ShowExpenseActivityUi : AnkoComponent<ShowExpenseActivity> {
    lateinit var toolbar : TemplateToolbar
    lateinit var name : TextView
    lateinit var date : TextView
    lateinit var value : TextView
    lateinit var showInOverview : TextView
    lateinit var chargeable : TextView
    lateinit var bill : TextView
    lateinit var chargeNextMonth : TableRow

    override fun createView(ui: AnkoContext<ShowExpenseActivity>) = with(ui) {
        verticalLayout {
            toolbar = toolbarTemplate {}

            tableViewFrame {
                tableRow {
                    static { text = resources.getString(R.string.name) }
                    name = staticWithData { id = R.id.act_show_expense_name }
                }
                tableRow {
                    static { text = resources.getString(R.string.date) }
                    date = staticWithData { id = R.id.act_show_expense_date }
                }
                tableRow {
                    static { text = resources.getString(R.string.income) }
                    value = staticWithData { id = R.id.act_show_expense_value }
                }
                tableRow {
                    static { text = resources.getString(R.string.income_to_show_in_overview) }
                    showInOverview = staticWithData { id = R.id.act_show_expense_value_to_show_in_overview }
                }
                tableRow {
                    static { text = resources.getString(R.string.paid_with) }
                    chargeable = staticWithData { id = R.id.act_show_expense_chargeable }
                }
                tableRow {
                    static { text = resources.getString(R.string.bill) }
                    bill = staticWithData { id = R.id.act_show_expense_bill }
                }
                chargeNextMonth = tableRow {
                    id = R.id.act_show_expense_charge_next_month
                    static { text = resources.getString(R.string.pay_next_month) }
                }
            }
        }.applyRecursively(::applyTemplateViewStyles)
    }
}
